### ISSUE TITLE
Kiwix-serve: external link blocker does not respect the `target` attribute

### ISSUE THREAD

**Issue by benoit74** [kiwix/libkiwix#1166]
See e.g. https://dev.library.kiwix.org/viewer#libretexts.org_en_workforce_2024-12/index.html (or any other libretexts ZIM in dev library ... or prod soon hopefully).

When you click on the blue globe in the top right corner, the target is `_blank` but the external link blocker opens in same tab.

![Image](https://github.com/user-attachments/assets/e665a009-dfe0-4d88-9ac5-068b02af9de4)

![Image](https://github.com/user-attachments/assets/8b139fa6-55d2-4b77-9c8e-3c55b3541a89)

**Comment by Optimus-NP** [comment]
Thanks to @veloman-yunkan 

✅ This issue has been resolved by [PR #1176](https://github.com/kiwix/libkiwix/pull/1176).

The problem was that the external link blocker was not respecting the _blank target, causing external links (like the blue globe icon in the top right) to open in the same tab instead of a new one.

Fix Details:

-  The fix ensures that links explicitly set to open in _blank are properly handled.
-  Now, clicking external links opens them in a new tab as expected.

**Comment by kelson42** [comment]
@Optimus-NP Thx, we will check and close

**Comment by kelson42** [comment]
First of all the feature to "block" loading of external pages is clear and documented, it is `--blockexternal`.

Therefore, the default is option is to not block.

Now, after merging #1176 I guess, the original bug is not there but we still have a bug!

Current behaviour is that it open the link in a new tab like the HTML code requests. But there is no blocking at all appearing if `--blockexternal` is specified. We should see the blocking page in a new tab, this is how it should behave.

Hope this helps @spano @jejbq @Optimus-NP @benoit74

**Comment by benoit74** [comment]
I think current behavior is fine. I suspect #1176 fixed this issue as well.

https://github.com/user-attachments/assets/5a4c8eb7-e84f-48ea-ade7-180cee35b9fd

**Comment by jejbq** [comment]
> First of all the feature to "block" loading of external pages is clear and documented, it is --blockexternal.
> Therefore, the default is option is to not block.

I don't understand how the default option not to block is implemented.

We are using [Zimit](https://github.com/openzim/zimit) and by default all external links are blocked except if the original link is using target=_blank. All our ZIMs are shared via [kiwix-serv](https://github.com/kiwix/kiwix-tools) and _blank bypasses the default block. We would prefer not to block, but we don't understand where the blocking is configured in kiwix-serv, libkiwix or ZIM produced by zimit.

**Comment by kelson42** [comment]
@benoit74 I don't understand why you close, I have clearly reported the wrong behaviour in my last comment.

**Comment by veloman-yunkan** [comment]
> Now, after merging [#1176](https://github.com/kiwix/libkiwix/pull/1176) I guess, the original bug is not there but we still have a bug!
> 
> Current behaviour is that it open the link in a new tab like the HTML code requests. But there is no blocking at all appearing if `--blockexternal` is specified. We should see the blocking page in a new tab, this is how it should behave.

@kelson42 I can't reproduce the claimed bug the way I understood it from your description using a [minimal test case](https://github.com/user-attachments/files/25621892/kiwix-tools_issue.678_test.zim.zip) created for kiwix/kiwix-tools#678.

Yes, external link blocking can be by-passed by opening external links in a new window through explicit user actions such as:
- middle-click on a link
- right-click on a link and select "Open link in a new tab/window" from the context menu
- right-click on a link, select "Copy link" from the context menu and use it a new tab/window

Those cases can also be fixed. I will do that.

**Comment by veloman-yunkan** [comment]
Using the [test ZIM file](https://github.com/user-attachments/files/25767506/testing_links_in_kiwix-serve_viewer.zim.zip) from #1281 I have come across the following current problems with external links:

- same-window javascript links to external resources are blocked by the browser due to content-security policy even when external link blocking by `kiwix-serve` is off (this is intended to restrict the scripts from escaping from the viewer sandbox).
- external link blocking by `kiwix-serve` can be circumvented via javascript links opening in a new window
- external link blocking by `kiwix-serve` can be circumvented by manually forcing a regular (non-JS) link to be opened in a new window (e.g. via a middle-button click, or context menu)

@kelson42 How should we prioritize those issues?

**Comment by kelson42** [comment]
@veloman-yunkan If you are already on it, I'm glad if you can nail it down. We introduced this frame based viewer 3 years ago, and would be glad if we could move on to the "next generation" of bugs!

**Comment by jejbq** [comment]
> external link blocking by kiwix-serve can be circumvented by manually forcing a regular (non-JS) link to be opened in a new window (e.g. via a middle-button click, or context menu)

You can also add `target="_blank"` to the `<a href=` link to bypass  the blocking.

My question is simple: 
**How do we disable any blocking of external URLs in kiwix-serve?**

In my point of view, the blocking is enabled by default (and we did not use --blockexternal).
We want to open it all to make the same behavior with or without target="_blank".

Thank you for your insight

**Comment by veloman-yunkan** [comment]
@jejbq I think that you mean a different kind of link blocking. Can you please share a small ZIM file with such behaviour?

**Comment by jejbq** [comment]
@veloman-yunkan

[test.zim.gz](https://github.com/user-attachments/files/25880808/test.zim.gz)

**Using Firefox**

First link (A) will display **Sorry, the url ... is not found on this server**
Second link (B) will redirect to the same URL as (A) because of target="_blank"

We would like (A) and (B) to redirect to the external URL without blocking it

Software Stack:
- kiwix-serve 3.8.2 (latest as of today) via Docker
``` yaml
  kiwix-serve:
    image: ghcr.io/kiwix/kiwix-serve:3.8.2
    hostname: kiwix-serve
    read_only: true
    command:
      - '--monitorLibrary'
      - '--library'
      - 'library.xml'
    volumes:
      - archive:/data:ro

volumes:
  archive:
```
- zimit:latest via Docker
``` bash
  docker run \
  --detach \
  --name "test" \
  --shm-size 2gb --memory 2g --cpu-shares 3072 \
  --volume archive:/output \
  ghcr.io/openzim/zimit:latest zimit \
  --workers 2 \
  --waitUntil domcontentloaded \
  --output="/output/test" \
  --statsFilename="/output/test/task_progress.json" \
  --verbose \
  --keep \
  --name="test" \
  --zim-file="test.zim" \
  --url="http://$(hostname -i)" \
  --title="Test" \
  --description="Test" \
  --favicon="https://wiki.kiwix.org/w/images/favicon.ico"

  kiwix-manage library.xml add test.zim
```
- Python Dummy Web Server
``` html
mkdir test
cd test
cat <<'EOF' > index.html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
    <title>Test</title>
  </head>
  <body class="section-portrait">
  <ul>
  <li><a href="http://www.ethz.ch/index_EN">(A) Swiss Federal Institute of Technology in Zurich</a></li>
  <li><a href="http://www.ethz.ch/index_EN" target="_blank">(B) Swiss Federal Institute of Technology in Zurich</a></li>
  </ul>
</html>
EOF

python3 -m http.server 80
```

**Comment by jejbq** [comment]
@veloman-yunkan 

Load [ethz.zim.gz](https://github.com/user-attachments/files/25883251/ethz.zim.gz) in kiwix-serve 3.8.2

- **(A) ETH Zurich 404 over HTTP**
- (B) ETH Zurich 404 + target _blank over HTTP
- **(C) ETH Zurich 404 over HTTPS**
- (D) ETH Zurich 404 + target _blank over HTTPS
- **(E) ETH Zurich over HTTP**
- (F) ETH Zurich + target _blank over HTTP
- **(G) ETH Zurich over HTTPS**
- (H) ETH Zurich + target _blank over HTTPS

**With Firefox**

**(A) will display**
**`Sorry, the url http://www.ethz.ch/index_EN is not found on this server`**

**(C) will display**
**`Sorry, the url https://www.ethz.ch/index_EN is not found on this server`**

**(E) will display**
**`Sorry, the url http://www.ethz.ch/ is not found on this server`**

**(G) will display**
**`Sorry, the url https://www.ethz.ch/ is not found on this server`**

**We want to avoid any **`Sorry, the url ... is not found on this server`** from kiwix-serve**

**P.S. Even `<a href="https://ethz.ch/en.html">(I) ETH Zurich Real (no 301)</a>` will not work with Firefox**
**`Sorry, the url https://ethz.ch/en.html is not found on this server`**

- **(I) ETH Zurich Real (no 301)**
- (J) ETH Zurich Real (no 301) + target _blank over HTTPS

``` html
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
    <meta http-equiv="Content-Type" content="text/html;charset=utf-8"/>
    <title>Test</title>
</head>
<body class="section-portrait">
  <ul>
  <li><a href="http://www.ethz.ch/index_EN">(A) ETH Zurich 404 over HTTP</a></li>
  <li><a href="http://www.ethz.ch/index_EN" target="_blank">(B) ETH Zurich 404 + target _blank over HTTP</a></li>
  <li><a href="https://www.ethz.ch/index_EN">(C) ETH Zurich 404 over HTTPS</a></li>
  <li><a href="https://www.ethz.ch/index_EN" target="_blank">(D) ETH Zurich 404 + target _blank over HTTPS</a></li>
  <li><a href="http://www.ethz.ch/">(E) ETH Zurich over HTTP</a></li>
  <li><a href="http://www.ethz.ch/" target="_blank">(F) ETH Zurich + target _blank over HTTP</a></li>
  <li><a href="https://www.ethz.ch/">(G) ETH Zurich over HTTPS</a></li>
  <li><a href="https://www.ethz.ch/" target="_blank">(H) ETH Zurich + target _blank over HTTPS</a></li>
  <li><a href="https://ethz.ch/en.html">(I) ETH Zurich Real (no 301)</a></li>
  <li><a href="https://ethz.ch/en.html" target="_blank">(J) ETH Zurich Real (no 301) + target _blank over HTTPS</a></li>
  <li><a href="http://kiwix.org/en/">(K) Kiwix over HTTP</a></li>
  <li><a href="https://kiwix.org/en/">(L) Kiwix over HTTPS</a></li>
  </ul>
  </body>
</html>
```

**Comment by veloman-yunkan** [comment]
@jejbq Thanks for the test data. Though definitely a bug on our side, it has nothing to do with link blocking and deserves a separate ticket.
